lib/connect/api/base.js

```diff
@@ -5,69 +5,210 @@
+    const options = {
+      method: 'PUT',
+      body: obj,
+    };
+    const response = await this.fetch(url, options);
+    return response.json();
+  }
+
+  async list(options = {}) {
+    return this.search(options);
+  }
+
+  /* eslint-disable max-len */
+
+  /**
+   * Search for a resource.
+   *
+   * @param  {Object}         options              The search options object as described below.
+   * @param  {string|Query}   options.query        The query to send to the server as a RQL string or RQL query object.
+   * @param  {string|Array}   options.orderBy      The ordering criteria.
+   * @param  {number}         [options.limit=100]  The maximum number of results to return.
+   * @param  {number}         [options.offset=0]   The offset from which to start to include results.
+   *
+   * @return {Array} An array of resourcesthat match the provided filters.
+   */
+  async search(options = {}) { /* eslint-enable max-len */
+    let url = this.baseUri;
+    url = this.addQuery(url, options.query);
+    url = this.addSorting(url, options.orderBy);
+    url = this.addPaging(url, options.limit, options.offset);
+    const response = await this.fetch(url);
+    return response.json();
+  }
+
+  /* eslint-disable class-methods-use-this */
 
-  addParams(url, params) {
+  /**
+   * Append parameters to the querystring.
+   *
+   * @param   {string}         url      The URL to which append the parameters.
+   * @param   {string|Object}  params   An plain object of paramers or a partial querystring.
+   *
+   * @return  {string}          The modified URL with querystring parameters.
+   */
+  appendToQuerystring(url, params) {
     let computedUrl = url;
     if (params) {
       const separator = url.includes('?') ? '&' : '?';
-      const qsParams = querystring.stringify(params);
-      if (qsParams) {
-        computedUrl += `${separator}${qsParams}`;
+      if (typeof params === 'string') {
+        computedUrl += `${separator}${params}`;
+      } else if (_.isPlainObject(params)) {
+        const qsParams = querystring.stringify(params);
+        if (qsParams) {
+          computedUrl += `${separator}${qsParams}`;
+        }
+      } else {
+        throw new Error('Only strings and plain objects are supported.');
       }
     }
     return computedUrl;
   }
 
   /**
-   * Check if the response is ok (status = 2xx)
+   * Add a RQL query to the querystring.
    *
-   * @param {Object} response An http response object.
+   * @param   {string}        url    The URL to which append the RQL query.
+   * @param   {string|Query}  query  The RQL query to append.
+   *
+   * @return  {string}         The modified URL with the RQL query included.
+   */
+  addQuery(url, query) {
+    if (query) {
+      if (typeof query === 'string' || query instanceof Query) {
+        const q = query.toString();
+        return this.appendToQuerystring(url, q);
+      }
+      throw new Error('Only rql string or rql Query are supported');
+    }
+    return url;
```

Comment:
Mutating passed argument directly is not really good.
Suppose one day we pass here not a String, but reference to an Object. Function returns value - so we would like to have some kind of condition outside - assign if A, not assign if B. Suppose we revealed B case and want to use previous value. We won't be able cause it will be mutated inside by link.
In general I would prefer to see this transformations outside of class scope as pure functions.

Reply:
I don't understand. This function get an url and a query and if query is of expected type it returns a computed url with the query appended to it, otherwise returns the unmodifed url. Why do you say that the argument is mutating ?

The idea behind this functions is that implementors can "override" the parent function to customize behaviours. For example for the RequestResource that does not support yet rql, we can override the addQuery method to transform a rql query to "flat" querystring arguments as per the endpoint specs. The appendToQuerystring method could be moved outside the class scope.